### ISSUE TITLE
Use installed googletests/gmcok as default, do not force installing

### ISSUE THREAD

**Issue by chivakker** [Intel-Media-SDK/MediaSDK#1165]
GoogleTests integration in MediaSDK is broken.

Do not modify filesystem with googletest installation.  Use pkgconfig gtest and pkgconfig gmock to find if system already supports.  If not then force user to enable it.  Install it in a particular prefix as you don't know if user wants to install and suggest installing it in a temp folder.

**Comment by dvrogozh** [comment]
Ok, I reproduced it with the following on CentOS 7:
```
# scl enable devtoolset-7 -- cmake -DCMAKE_INSTALL_PREFIX=`pwd`/install -DBUILD_TESTS=ON ..
# make
# make install
# tree `pwd`/install
install
├── include
│   ├── gmock
│   │   ├── gmock-actions.h
│   │   ├── gmock-cardinalities.h
│   │   ├── gmock-generated-actions.h
│   │   ├── gmock-generated-actions.h.pump
│   │   ├── gmock-generated-function-mockers.h
│   │   ├── gmock-generated-function-mockers.h.pump
│   │   ├── gmock-generated-matchers.h
│   │   ├── gmock-generated-matchers.h.pump
│   │   ├── gmock-generated-nice-strict.h
│   │   ├── gmock-generated-nice-strict.h.pump
│   │   ├── gmock.h
│   │   ├── gmock-matchers.h
│   │   ├── gmock-more-actions.h
│   │   ├── gmock-more-matchers.h
│   │   ├── gmock-spec-builders.h
│   │   └── internal
│   │       ├── custom
│   │       │   ├── gmock-generated-actions.h
│   │       │   ├── gmock-generated-actions.h.pump
│   │       │   ├── gmock-matchers.h
│   │       │   ├── gmock-port.h
│   │       │   └── README.md
│   │       ├── gmock-generated-internal-utils.h
│   │       ├── gmock-generated-internal-utils.h.pump
│   │       ├── gmock-internal-utils.h
│   │       └── gmock-port.h
│   ├── gtest
│   │   ├── gtest-death-test.h
│   │   ├── gtest.h
│   │   ├── gtest-message.h
│   │   ├── gtest-param-test.h
│   │   ├── gtest-param-test.h.pump
│   │   ├── gtest_pred_impl.h
│   │   ├── gtest-printers.h
│   │   ├── gtest_prod.h
│   │   ├── gtest-spi.h
│   │   ├── gtest-test-part.h
│   │   ├── gtest-typed-test.h
│   │   └── internal
│   │       ├── custom
│   │       │   ├── gtest.h
│   │       │   ├── gtest-port.h
│   │       │   ├── gtest-printers.h
│   │       │   └── README.md
│   │       ├── gtest-death-test-internal.h
│   │       ├── gtest-filepath.h
│   │       ├── gtest-internal.h
│   │       ├── gtest-linked_ptr.h
│   │       ├── gtest-param-util-generated.h
│   │       ├── gtest-param-util-generated.h.pump
│   │       ├── gtest-param-util.h
│   │       ├── gtest-port-arch.h
│   │       ├── gtest-port.h
│   │       ├── gtest-string.h
│   │       ├── gtest-tuple.h
│   │       ├── gtest-tuple.h.pump
│   │       ├── gtest-type-util.h
│   │       └── gtest-type-util.h.pump
│   └── mfx
│       ├── mfxastructures.h
│       ├── mfxaudio.h
│       ├── mfxaudio++.h
│       ├── mfxbrc.h
│       ├── mfxcamera.h
│       ├── mfxcommon.h
│       ├── mfxdefs.h
│       ├── mfxdispatcherprefixedfunctions.h
│       ├── mfxenc.h
│       ├── mfxfei.h
│       ├── mfxfeihevc.h
│       ├── mfxjpeg.h
│       ├── mfxla.h
│       ├── mfxmvc.h
│       ├── mfxpak.h
│       ├── mfxplugin.h
│       ├── mfxplugin++.h
│       ├── mfxscd.h
│       ├── mfxsc.h
│       ├── mfxsession.h
│       ├── mfxstructures.h
│       ├── mfxvideo.h
│       ├── mfxvideo++.h
│       ├── mfxvp8.h
│       ├── mfxvp9.h
│       └── mfxvstructures.h
├── lib64
│   ├── cmake
│   │   └── GTest
│   │       ├── GTestConfig.cmake
│   │       ├── GTestConfigVersion.cmake
│   │       ├── GTestTargets.cmake
│   │       └── GTestTargets-release.cmake
│   ├── libgmock.a
│   ├── libgmock_main.a
│   ├── libgtest.a
│   ├── libgtest_main.a
│   ├── libmfxhw64.so -> libmfxhw64.so.1
│   ├── libmfxhw64.so.1 -> libmfxhw64.so.1.28
│   ├── libmfxhw64.so.1.28
│   ├── libmfx.so -> libmfx.so.1
│   ├── libmfx.so.1 -> libmfx.so.1.28
│   ├── libmfx.so.1.28
│   ├── mfx
│   │   ├── libmfx_h264la_hw64.so
│   │   ├── libmfx_hevcd_hw64.so
│   │   ├── libmfx_hevce_hw64.so
│   │   ├── libmfx_hevc_fei_hw64.so
│   │   ├── libmfx_vp8d_hw64.so
│   │   ├── libmfx_vp9d_hw64.so
│   │   └── libmfx_vp9e_hw64.so
│   └── pkgconfig
│       ├── gmock_main.pc
│       ├── gmock.pc
│       ├── gtest_main.pc
│       ├── gtest.pc
│       ├── libmfxhw64.pc
│       ├── libmfx.pc
│       └── mfx.pc -> libmfx.pc
└── share
    └── mfx
        ├── plugins.cfg
        ├── samples
        │   ├── libcttmetrics.so
        │   ├── libsample_plugin_opencl.so
        │   ├── libsample_rotate_plugin.so
        │   ├── libvpp_plugin.a
        │   ├── metrics_monitor
        │   ├── ocl_rotate.cl
        │   ├── sample_decode
        │   ├── sample_encode
        │   ├── sample_fei
        │   ├── sample_hevc_fei
        │   ├── sample_hevc_fei_abr
        │   ├── sample_multi_transcode
        │   ├── sample_vpp
        │   └── test_monitor
        └── tests
            └── mfx_dispatch_test

17 directories, 123 files
```

@chivakker : please, avoid using -DBUILD_TESTS=ON or -DBUILD_ALL=ON to workaround the problem.

**Comment by dvrogozh** [comment]
@vshampor : can you, please, look into this? This is critical. Why this line did not work as intended:
https://github.com/Intel-Media-SDK/MediaSDK/blob/d8fe7458633b17d556a35a4d7e5f570e450a9e2f/CMakeLists.txt#L115

**Comment by chivakker** [comment]
@dvrogozh the problem is more complex than that.  In the test folder you guys have cttmonitor_gtest which unconditionally assumes gtest is installed per your local copy of gtest.  MediaSDK is not checking for system installed gtest, as a  result your liking it as libgtest.so instead of doing something like -lgtest as a pkg-config file will do.

Then when trying to package this more problems will pop up.  A ClearLinux rpm doesn't know where to get libgtest.so from, then dependencies are broken. 

Of course I adding to my cmake -DINSTALL_GTEST=OFF -DBUILD_GMOCK=OFF

**Comment by vshampor** [comment]
Greetings, @chivakker !

There is no mention of "cttmonitor" in the entire MediaSDK repo at the master branch. Could you please elaborate on where exactly you observed it and what actions reproduce further issues?

Otherwise, the PR #1166 fixes the issue at hand.

**Comment by dvrogozh** [comment]
@chivakker : please,  provide your _full_ cmake config line which you have used.

**Comment by chivakker** [comment]
@vshampor 

Let me try again with more detail

PR #1166 fixes things when user doesn't disable INSTALL_GTEST during configure (I don't know if that's the right terminology for cmake).  But there are other problems  that are not covered by this PR.

For example my cmake is exactly this

`cmake -DENABLE_WAYLAND=true -DENABLE_X11=true -DBUILD_TESTS=ON -DINSTALL_GTEST=OFF -DBUILD_MOCK=OFF

I want to have a test target, i.e. make target should do something useful.  

This test, mediasdk/samples/metrics_monitor/test/cttmetrics_gtest.cpp will compile assuming I have gtest-devel installed in my system.  For this reason, MediaSDK should check for pkg-config gtest to include it in your LIBS and INCLUDES directives inside  `mediasdk/samples/metrics_monitor/CMakeLists.txt `in the same way you're doing for LIBDRM.  I don't know why this is checked against PKG_PCIACCESS but neither LIBS or INCLUDES are used in your test_monitor executable

Then this

 `target_link_libraries( test_monitor pthread rt gtest cttmetrics ${PKG_LIBDRM_LIBRARIES})
`
From the expanded make command. 

`/usr/lib64/ccache/bin/g++  -msse4.2 -std=c++11 -m64 -g -O2 -fPIC -g -feliminate-unused-debug-types -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=32 -Wformat -Wformat-security -Wl,-z,now,-z,relro,-z,max-page-size=0x1000,-z,separate-code -Wno-error -ftree-vectorize -ftree-slp-vectorize -Wl,--enable-new-dtags -m64 -march=westmere -mtune=haswell -fasynchronous-unwind-tables -fno-omit-frame-pointer -Wp,-D_REENTRANT -pipe -fPIC -Wall -Wformat -Wformat-security -Wnon-virtual-dtor -Wno-deprecated-declarations -Wno-unknown-pragmas -Wno-unused -O2 -g -DNDEBUG  -fPIE -pie  CMakeFiles/test_monitor.dir/test/cttmetrics_gtest.cpp.o CMakeFiles/test_monitor.dir/test/igt_load.cpp.o  -o ../../__bin/RelWithDebInfo/test_monitor -lpthread -lrt ../../googletest/googlemock/gtest/libgtest.so ../../__bin/RelWithDebInfo/libcttmetrics.so -ldrm -pthread -Xlinker --start-group -Xlinker --end-group -lgcc`

This is linking against libgtest.so in your own folder, this should not be happening but instead it should be -lgtest if pkg-config was used.

**Comment by dvrogozh** [comment]
Reopening the issue since it was automatically closed by github on "Fixes: #1165" in PR #1166,  but we not yet finished discussion.

**Comment by dvrogozh** [comment]
@chivakker : I believe there is something else you specify either in the cmake configuration command line or thru some other cmake configuration channel. See the following build on CentOS 7:

```
# scl enable devtoolset-7 -- cmake -DENABLE_WAYLAND=true -DENABLE_X11=true -DBUILD_TESTS=ON -DINSTALL_GTEST=OFF -DBUILD_MOCK=OFF ..

...
CMake Warning:
  Manually-specified variables were not used by the project:
    BUILD_MOCK
...
# make
# ldd __bin/release/test_monitor
        linux-vdso.so.1 =>  (0x00007ffcdaf4d000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1d7791c000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f1d77714000)
        libcttmetrics.so => not found
        libdrm.so.2 => /lib64/libdrm.so.2 (0x00007f1d77503000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f1d771fc000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f1d76efa000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f1d76ce4000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f1d76917000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f1d77d96000)

# ldd __bin/release/mfx_dispatch_test
        linux-vdso.so.1 =>  (0x00007fff66cf9000)
        libmfx.so.1 => not found
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f96038ea000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f96036e6000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f96033df000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f96030dd000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f9602ec7000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f9602afa000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f9603de5000)

```

And some grep for linker:
`samples/metrics_monitor/CMakeFiles/test_monitor.dir/link.txt:1:/opt/rh/devtoolset-7/root/usr/bin/c++  -msse4.2 -std=c++11 -m64 -g  -pipe -fPIC -Wall -Wformat -Wformat-security -Wnon-virtual-dtor -Wno-deprecated-declarations -Wno-unknown-pragmas -Wno-unused -O2 -DNDEBUG  -fPIE -pie  CMakeFiles/test_monitor.dir/test/cttmetrics_gtest.cpp.o CMakeFiles/test_monitor.dir/test/igt_load.cpp.o  -o ../../__bin/release/test_monitor -lpthread -lrt ../../googletest/googlemock/gtest/libgtest.a ../../__bin/release/libcttmetrics.so -ldrm -pthread -Xlinker --start-group -Xlinker --end-group -lgcc
`

So:
1. Don't specify BUILD_MOCK - this looks useless
2. Your build for some reason picks shared gtest library while we were supposed to pick static library during within mediasdk build. @chivakker : can you, please, try to check what exactly in your build ecosystem was responsible for that?

**Comment by chivakker** [comment]
Ok, I'm able to duplicate the problem on Fedora.  These are my observations. 

I am also forcing to build shared libraries, this is my cmake

cmake -DENABLE_WAYLAND=true -DENABLE_X11=true -DBUILD_TESTS=ON -DINSTALL_GTEST=OFF -DBUILD_MOCK=OFF -DBUILD_SHARED_LIBS:BOOL=ON

which in turn builds googletest dynamically from mediaSDK sources producing libgtest.so which is used during linking of test_monitor.  A similar thing happens with libgmock.so and mfx_dispatch_test.

That's all fine till I run test_monitor as nobody installed libgtest.so.

If I force googletest to build as type=STATIC then everything works as you guys originally tested.

So, can you guys make sure that this is fixed?  pkg-config should be used and linking has to happen against system libgtest.so.1.8.1 as in fedora.  This is also hinting that you need to support the following flags

BUILD_GMOCK default NO
INSTALL_GMOCK default NO
BUILD_GTEST default NO
INSTALL_GTEST default NO

**Comment by dvrogozh** [comment]
gtest is pretty specific thing and usually is an exception for the projects to use it from the system dependency. Usually what projects do is coping gtest which is built and used. So, we just tried to follow what others are doing with the library. For example, that's what https://github.com/intel/libva-utils/tree/master/test/gtest is doing. I wonder how that ruled out system gtest installation.

Anyhow, that's an issue which I am not yet sure how to fix. We will investigate. In a meanwhile, please, avoid BUILD_TESTS=ON option. It won't give you much anyway as of now, though we definitely plan to bring it alive and widely extend.

**Comment by chivakker** [comment]
Another note:

Consider also avoiding to install mfx_dispatch_test and test_monitor and just run both with make test which should figure out where libgtest.so is installed within your package.

This should be enough to enable tests without borking systems.

**Comment by vshampor** [comment]
@chivakker, if we wanted our tests to link against system gtest/gmock, we wouldn't have included the entire gtest/gmock source in our repo. Rather, we want our tests to use a controlled version of the gtest/gmock. 

Also, perhaps I am missing some major use case/requirement here, but why would anyone want to use a shared library for such an auxiliary and compact thing as gtest/gmock, when it works perfectly fine while linked statically (by default)? The reason we run the tests is to validate the MSDK code. The statically linked gtest/gmock code, as a supporting framework, does a good job of it right now - we don't need to leverage improvements (if any) of this framework from a system-managed shared library version of gtest/gmock.

**Comment by dvrogozh** [comment]
@vshampor : -DBUILD_SHARED_LIBS:BOOL=ON is a standard widely used flag: https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html. The intent is to make sure that you generate .so libraries for those cases when project is capable to provide both .so and .a. Our problems, I believe, is that we did not fully configured our build of gtest and did not force to use .a only. As a result the build reacted on the above flag and went wrong. I think that's the major issue right now. Can you, please, check how we can avoid that? Probably we need to set initially BUILD_SHARED_LIBS=OFF since I don't think mediasdk itself provides support for this flag at all. When resolved, let's think about installability of the tests.

**Comment by vshampor** [comment]
@dvrogozh, so this is a standard flag, which is shared across all levels of CMakeLists.txt. Well, the user gets exactly what he asked for here - all libraries are built as shared. If this was not the intent, why specify this flag at all? Do we still have any static library build targets with an open API in this repo?

To address the question whether we can force the flag to OFF only for gtest/gmock - we can, observe: #1179

**Comment by chivakker** [comment]
PR #1179 should be completed by not installing the applications linked against googletest.  You have said they are intended for build purposes only, so don't install them
